expression/util.go

```diff
@@ -273,3 +274,104 @@ func Contains(exprs []Expression, e Expression) bool {
 	}
 	return false
 }
+
+// ExtractFiltersFromDNFs checks whether the cond is DNF and extract the same expression in its leaf items.
+func ExtractFiltersFromDNFs(ctx context.Context, conditions []Expression) []Expression {
+	var allExtracted []Expression
+	for i := len(conditions) - 1; i >= 0; i-- {
+		if sf, ok := conditions[i].(*ScalarFunction); ok && sf.FuncName.L == ast.LogicOr {
+			extracted, remained := extractFiltersFromDNF(ctx, sf)
+			allExtracted = append(allExtracted, extracted...)
+			if remained == nil {
+				conditions = append(conditions[:i], conditions[i+1:]...)
+			} else {
+				conditions[i] = remained
+			}
+		}
+	}
+	return append(conditions, allExtracted...)
+}
+
+// extractFiltersFromDNF extracts the same condition that occurs in every DNF item.
+func extractFiltersFromDNF(ctx context.Context, dnfFunc *ScalarFunction) ([]Expression, Expression) {
+	dnfItems := FlattenDNFConditions(dnfFunc)
+	sc := ctx.GetSessionVars().StmtCtx
+	codeMap := make(map[string]int)
+	hashcode2Expr := make(map[string]Expression)
+	firstRun := true
+	for _, dnfItem := range dnfItems {
+		// We need this because there may be the case like `select * from t, t1 where (t.a=t1.a and t.a=t1.a) or (something).
+		// We should make sure that the two `t.a=t1.a` contributes only once.
+		innerMap := make(map[string]struct{})
+		if sf, ok := dnfItem.(*ScalarFunction); ok && sf.FuncName.L == ast.LogicAnd {
```

Comment:
We don't need to check if it is a real cnf. It's all right to flatten conditions directly and loop it.